packages/mattermost-redux/src/selectors/createSelector.js

```diff
@@ -0,0 +1,155 @@
+        const id = generateId();
+        let recomputations = 0;
+        let calls = 0;
+        const resultFunc = funcs.pop();
+        const dependencies = getDependencies(funcs);
+
+        const memoizedResultFunc = memoize(
+            function() {
+                recomputations++
+                // apply arguments instead of spreading for performance.
+                return resultFunc.apply(null, arguments)
+            },
+            null,
+            ...memoizeOptions
+        );
+
+        // If a selector is called with the exact same arguments we don't need to traverse our dependencies again.
+        const selector = memoize(function () {
+            const params = [];
+            const length = dependencies.length;
+
+            for (let i = 0; i < length; i++) {
+                // apply arguments instead of spreading and mutate a local list of params for performance.
+                params.push(dependencies[i].apply(null, arguments));
+            }
+
+            // apply arguments instead of spreading for performance.
+            return memoizedResultFunc.apply(null, params);
+        },
+        function() {
+            calls++
+        });
+
+        selector.resultFunc = resultFunc;
+        selector.dependencies = dependencies;
+        selector.recomputations = () => recomputations;
+        selector.resetRecomputations = () => recomputations = 0;
+        selector.calls = () => calls;
+        selector.resetCalls = () => calls = 0;
+        selector.effectiveness = () => 100 - ((recomputations / calls) * 100);
+        selector.selectorName = () => name;
+        selector.id = () => id;
+        trackedSelectors[id] = selector;
+        return selector;
+    };
+}
+
+// resetTrackedSelectors resets all the measurements for memoization effectiveness.
+function resetTrackedSelectors() {
+    Object.values(trackedSelectors).forEach((selector) => {
+        selector.resetCalls();
+        selector.resetRecomputations();
+    });
+}
+
+// getSortedTrackedSelectors returns an array, sorted by effectivness, containing mesaurement data on all tracked selectors.
+function getSortedTrackedSelectors() {
+    let selectors = Object.values(trackedSelectors);
+    selectors = selectors.map((selector) => ({name: selector.selectorName(), effectiveness: selector.effectiveness(), recomputations: selector.recomputations(), calls: selector.calls()}));
+    selectors.sort((a, b) => (a.effectiveness > b.effectiveness) ? 1 : ((b.effectiveness > a.effectiveness) ? -1 : 0))
+    return selectors;
+}
+
+// dumpTrackedSelectorsStatistics prints to console a table containing the measurement data on all tracked selectors.
+function dumpTrackedSelectorsStatistics() {
+    const selectors = getSortedTrackedSelectors();
+    console.table(selectors); //eslint-disable-line no-console
+}
+
+// Only track selectors in development builds.
+if (process.env.NODE_ENV === 'development') { //eslint-disable-line no-process-env
```

Comment:
should there be a flag other than development mode to enable this?

Reply:
Probably, yeah